### DISCUSSION TITLE
I no longer can get Homebrew to function, zsh: command not found: brew

### DISCUSSION THREAD

**Discussion by OldSubSailor** [Homebrew/discussions#5259] [Getting started]
### Output of `brew config`

```shell
Can not get that far
```

### Output of `brew doctor`

```shell
Can not get that far
```

### Description of issue

I just got a new Mac mini (M2) and Sonoma 14.4.1. I had no issues getting Homebrew to function after installing it on my old iMac and Ventura and previous versions of macOS/osX
If I use the recommended:

echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> /Users/me/.zprofile
eval "$(/opt/homebrew/bin/brew shellenv)"

I can get brew commands to work for THAT terminal session only. I just discovered that I can only get things that were installed by Homebrew such as Python3.11 to work during that same terminal session **only** also.  
Therefore, once I exit the terminal and quit it,  all things Homebrew are gone when I reopen terminal and I have to go through entering those commands again. 
When I have brew working I get no negative reports from brew doctor. 

Is there a real fix for this?

**Comment by SMillerDev** [comment]
What does `cat ~/.zprofile` say regarding Homebrew?

**Comment by OldSubSailor** [comment]
You can see an entry for each time that I have entered the commands to pull “activate” Homebrew. 
However, there is no ORIGINAL version in .zprofile.pysave if that means anything. Just something for Python 3.8 and 3.9.




# Setting PATH for Python 3.11
# The original version is saved in .zprofile.pysave
PATH="/Library/Frameworks/Python.framework/Versions/3.11/bin:${PATH}"
export PATH



eval $(/opt/homebrew/bin/brew shellenv)
eval $(/opt/homebrew/bin/brew shellenv)
eval $(/opt/homebrew/bin/brew shellenv)
eval "$(/opt/homebrew/bin/brew shellenv)"
eval $(/opt/homebrew/bin/brew shellenv)
eval "$(/opt/homebrew/bin/brew shellenv)"
eval $(/opt/homebrew/bin/brew shellenv)
eval "$(/opt/homebrew/bin/brew shellenv)"
eval $(/opt/homebrew/bin/brew shellenv)
eval "$(/opt/homebrew/bin/brew shellenv)"

**Comment by OldSubSailor** [comment]
So does this need to become a brew issue, it feels like brew is broken.

**Reply by SMillerDev** [reply]
Brew is fine, you're just in a curious situation where your shell environment is not working.

What happens if you reduce all those eval lines in `~/.zprofile` to just one and then reboot your Mac. If it doesn't work after, let me know what `echo $PATH` says.

**Comment by OldSubSailor** [comment]
I do not know why it would have stopped working suddenly. It continues to work on Ventura.

As per your request:  

% bash list
bash: list: No such file or directory
% echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin

**Comment by OldSubSailor** [comment]
I also tried changing the path directly:

export PATH=$PATH/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/opt/homebrew/bin/

And entered it into .zshrc; however; it has no effect.

But it does as a direct command line entry from terminal (one session only) just like the other commands.

**Reply by SMillerDev** [reply]
What does `echo $SHELL` say?

**Comment by OldSubSailor** [comment]
This is a brand new Mac Mini, M2.
So as expected is reads:
% echo $SHELL
/bin/zsh

macOS Sonoma Version 14.4.1
-----------------------------------------------------------------------------------
I do not know if it matters, but .zprofile.pysave has no "# Setting PATH for Python . . . etc" for Python 3.11 as it does for versions 3.8 & 3.9.  nor any thing else reflecting current activity. 

I thought about taking this to the Apple Support Community, but the first thing that they will do is jump me for having Homebrew on the box.

**Reply by SMillerDev** [reply]
> I do not know if it matters, but .zprofile.pysave has no "# Setting PATH for Python . . . etc"

The `.zprofile.pysave` has no impact on anything. What does the `.zprofile` say?

**Reply by OldSubSailor** [reply]
It is a blank page now.

**Reply by SMillerDev** [reply]
It needs to have exactly one `eval "$(/opt/homebrew/bin/brew shellenv)"`

**Reply by OldSubSailor** [reply]
**Why?** given all that I have pointed out in the past?

That did not work at all in the past efforts as indicated above.  I started out with one in .zprofile and it would last ONE terminal session ONLY. Then I would have to re-enter the two command lines to get things to work again and another eval line would be entered into .zprofile and so on, hence the multiple listings that I showed you.  So in my opinion, currently it makes no difference if anything is in there or not. 

The path variables that I show below works for now with nothing entered into it although I suspect that I can enter the command lines to get it re-entered into .zprofile but it will do nothing for me, seeing as how the exports in .zshrc works. 

Again, I can ask the Apple Communities (is it a combination of an **M2** machine[actually?] and Sonoma?) about this, but the first thing that they will do is jump me about having Homebrew installed and blame all my issues on Homebrew; I have no doubt that someone will even say "Homebrew strikes again!"

**Reply by SMillerDev** [reply]
> Why? given all that I have pointed out in the past?

Because a quick search here and in the old support forum reveals that you are the very first person to ever have this issue. Millions of people use Homebrew, and a good chunk of those use M-series processors without this issue.

> Again, I can ask the Apple Communities (is it a combination of an M2 machine[actually?] and Sonoma?) about this, but the first thing that they will do is jump me about having Homebrew installed

Maybe the community does, but that's not something Homebrew can help with. Apple has featured Homebrew in their keynotes so clearly it's not the company position.

**Reply by OldSubSailor** [reply]
I have been installing Homebrew for years on various Macs and never had an issue until now. This is a new, clean M2 machine. I have uninstalled and reinstalled Homebrew several times and tried that command line several times and it has not worked except for one terminal session sittings.

**Reply by OldSubSailor** [reply]
Now this is what my path looks like now:
% echo $PATH
/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin
%

**Comment by OldSubSailor** [comment]
I finally got something to work, but I do not feel that it is a true long term solution. It took two path change edits to .zshrc:

export PATH=/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
export PATH="/opt/homebrew/sbin:$PATH"

**Comment by gromgit** [comment]
> That did not work at all in the past efforts as indicated above. I started out with one in .zprofile and it would last ONE terminal session ONLY. Then I would have to re-enter the two command lines to get things to work again and another eval line would be entered into .zprofile and so on, hence the multiple listings that I showed you. So in my opinion, currently it makes no difference if anything is in there or not.

I'd bet you're actually _resetting_ your `PATH` _after_ you `eval "$(/opt/homebrew/bin/brew shellenv)"`. That would also explain why adding Homebrew paths in your `.zshrc` seems to work, because that file is sourced _after_ `.zprofile`.

Run the following command:
```
zsh -xli </dev/null 2>&1 | tee /tmp/trace.log
```
then post the contents of `/tmp/trace.log` in this thread.

**Reply by OldSubSailor** [reply]
I am sorry, however, I dont think so. I had nothing there for a long time and it was not working until I ran the export paths; also, the echo paths command continued to show a very simple path line until I entered the export paths i.e. /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
now it shows a very complicated path that the Apple Community claims that the M2 Mac is supposed to have.

**Reply by OldSubSailor** [reply]
@gromgit 
Given that the paths have been created by the export commands and that Homebrew is now working, what will command that you are asking do or prove now?  It sounds like closing the barn door after the horse got out.

**Reply by gromgit** [reply]
> I am sorry, however, I dont think so.

Then I'm sorry too...to have to tell you that your (possibly faulty) recollection means diddly-squat when trying to resolve this particular problem, and that only seeing what's _really_ happening matters. Since you seem to have rejected my suggestion, you have just one option left to get to the bottom of this. It's a rather involved process:

1. `man zsh` and read the *STARTUP/SHUTDOWN FILES* section, taking careful note of the series of startup files that `zsh` will read, and in what order.
2. Manually go through each of those files in order, to figure out exactly what your `zsh` does when you start a new session. If one of those files sources another file, be sure to read that new file as well. Rinse and repeat, until you find out what's buggering your PATH, or give up in exhaustion.

But if you're tired of trying random stuff in the hopes that something will stick, and just want to find out what's happening on your machine that's screwing up your Homebrew config, simply run the command I posted, which actually _traces your entire shell startup process_ (hence the "trace" in the log file name). If you need help figuring it out, post it here like I suggested.

**TCIY**: The Choice Is Yours.

**Reply by OldSubSailor** [reply]
All you needed to say is what it would do with no trail of tears.

**Comment by OldSubSailor** [comment]
Here is something that I forgot about. When I set up this Mac mini, I did not use the new computer option, but used the copy from another Mac, in this case a 2017 intel iMac. So in reality I suppose that it was not really a new CLEAN mini.

**Comment by gromgit** [comment]
> @gromgit Given that the paths have been created by the export commands and that Homebrew is now working, what will command that you are asking do or prove now? It sounds like closing the barn door after the horse got out.

More like walking away from a duct-taped pinhole in a pressure vessel. The underlying problem is still there, but you're assuming it won't affect anything else now and forever.

**Reply by OldSubSailor** [reply]
I told someone else that I was uncomfortable with it. It just felt like a temp. fix.

**Reply by OldSubSailor** [reply]
What about the copy from another older Mac rather than a new clean setup?  I posted that path 3 days ago

**Reply by gromgit** [reply]
> I told someone else that I was uncomfortable with it. It just felt like a temp. fix.

Not only is it a temporary fix, it's the _wrong_ fix, because you've also removed a whole bunch of other directories from PATH, including the ones that, as you said, "the Apple Community claims that the M2 Mac is supposed to have."

**Reply by OldSubSailor** [reply]
That was the original path after set up.
now 
% echo $PATH
/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin
%

**Reply by OldSubSailor** [reply]
Does that temp log command keep running until I kill it?

**Reply by gromgit** [reply]
Nope, it stops automatically at the end of the startup process. Does your normal session startup require you to key in anything?

**Reply by OldSubSailor** [reply]
It appears to stop, however, at every command line, it seems to restart and leave a lot of lines and ends partly ends until I hit return,
leaving a list like this. 
update_terminal_cwd:11> ++i
+update_terminal_cwd:11> i <= 4
+update_terminal_cwd:22> printf '\e]7;%s\a' file://Mac-mini.local/tmp

**Reply by OldSubSailor** [reply]
I guess that it is printing the log to the screen also

**Reply by OldSubSailor** [reply]
the log is below

**Reply by gromgit** [reply]
> I guess that it is printing the log to the screen also

Yes, so that you can see the startup process in action. That bit about "restart and leave a lot of lines", however, suggests that you're actually still in the traced shell session. Type `exit` and hit Enter to close out that session, and you should stop seeing all the "update_terminal_cwd" business.

**Comment by OldSubSailor** [comment]
[trace.log](https://github.com/Homebrew/discussions/files/14906118/trace.log)

**Reply by OldSubSailor** [reply]
Hope that I did this right.

**Reply by gromgit** [reply]
Looks OK. In fact, it looks as if you're _not_ setting PATH manually in `.zshrc` [as you said two days ago](https://github.com/orgs/Homebrew/discussions/5259?notification_referrer_id=NT_kwDOAANKlrIxMDEzMTYzMjM1ODoyMTU3MDI&notifications_query=repo%3AHomebrew%2Fdiscussions#discussioncomment-9029726). Did you remove those bits?

**Reply by gromgit** [reply]
And do you now only have one instance of `eval "$(/opt/homebrew/bin/brew shellenv)"` in your `.zprofile`, and nothing else?

**Reply by OldSubSailor** [reply]
Yeah, I went in with bbedit and changed those files. Actually, I just wiped the internal SSD and reinstalled Sonoma and now the path I suppose is 'normal' to start with:
/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin

but for some reason I forgot to copy my .zshrc file to my external drive. I can't believe it. However, I have one up in East Texas that I email myself in a week or two. 

but I guess that it is reinstall Homebrew and see what happens now.  Clouded over, so no eclipse for me.

**Reply by OldSubSailor** [reply]
That got it. Homebrew is working. That transfer from that old iMac somehow appears to have gotten Ventura and Sonoma all mixed together.
